### PR TITLE
Optimal SNS

### PR DESCRIPTION
**Pull request by rethink-forrest**
Copies over the functionality of the Optimal SNS velocity IK solver. OSNSsmVelocityIK includes the scale margin factor.

### CODE REVIEW

File: sns_ik_lib/src/osns_sm_velocity_ik.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <sns_ik/osns_sm_velocity_ik.hpp>
+
+using namespace Eigen;
+using namespace sns_ik;
+
+OSNSsmVelocityIK::OSNSsmVelocityIK(int dof, Scalar loop_period) :
```
Comment (rethink-imcmahon) [inline sns_ik_lib/src/osns_sm_velocity_ik.cpp]: The "sm" here and in the filename is non-obvious. Could you Expand that a bit so it's readable?

File: sns_ik_lib/src/sns_velocity_ik.cpp [hunk 2]
```diff
@@ -72,15 +72,10 @@ void SNSVelocityIK::setNumberOfTasks(int ntasks, int dof)
...
-    W.clear();
-    scaleFactors.clear();
-    dotQopt.clear();
-
-    for (int i = 0; i < n_tasks; i++) {
-      W.push_back(I);
-      scaleFactors.push_back(scale);
-      dotQopt.push_back(dq);
-    }
+    W.resize(n_tasks, I);
```
Comment (rethink-imcmahon) [inline sns_ik_lib/src/sns_velocity_ik.cpp]: +1. I like that this change takes the conditional re-sizing out of the mix.

**Comment by rethink-imcmahon** [comment]
+1 thank you for breaking all of this functionality out into classes. I just had one comment on a file / class name.

File: sns_ik_lib/src/osns_sm_velocity_ik.cpp (hunk 1 above)
Comment (rethink-forrest) [inline sns_ik_lib/src/osns_sm_velocity_ik.cpp]: sm == scale margin

Expanding the acronym would expand the file name:
osns_sm_velocity_ik.hpp -> osns_scaleMargin_velocity_ik.hpp

and class name:
OSNSsmVelocityIK -> OSNSScaleMarginVelocityIK

Without context, does the much longer name make things any clearer?

Comment (rethink-forrest) [inline sns_ik_lib/src/osns_sm_velocity_ik.cpp]: How about just renaming the class OSNS_sm_VelocityIK and leaving the file name the same?

**Comment by rethink-forrest** [comment]
I renamed to class for readability. There's a general open ticket to add better documentation to all of the classes.

**Comment by rethink-imcmahon** [comment]
Looks good!

**Pull request merged**